### ISSUE TITLE
In CIFAR10, why the fine-tuning training accuracy (<90%) is much lower than test accuracy (96.6%)?

### ISSUE THREAD

**Issue by chaoyanghe** [google-research/vision_transformer#30]

**Comment by andsteing** [comment]
This is probably due to the data augmentation. Check out the "augmented" training data:

https://colab.sandbox.google.com/github/google-research/vision_transformer/blob/master/vit_jax.ipynb#scrollTo=jFqi3h7yMEsB

You can see that those images actually are a lot harder to recognize and thus the training accuracy is lower. Training on these still helps the network to score better on the testset though. You can modify the Colab to train without data augmentation, and you should see how the train accuracy is better than the test accuracy.

**Comment by chaoyanghe** [comment]
@andsteing but why there isn't such a phenomenon when training with ResNet/MobileNet which also uses data augmentation?

**Comment by chaoyanghe** [comment]
I found when I use 0.5 mean and std as follows, the test accuracy can be higher (96%) and the training accuracy is super low (82%), but when I change it to the real mean and std, the test accuracy drops to 93% but the training accuracy is good at 95%. 

    # CIFAR_MEAN = [0.49139968, 0.48215827, 0.44653124]
    # CIFAR_STD = [0.24703233, 0.24348505, 0.26158768]

    CIFAR_MEAN = [0.5, 0.5, 0.5]
    CIFAR_STD = [0.5, 0.5, 0.5]

**Comment by andsteing** [comment]
Did you use ResNet/MobileNet on the same input pipeline? (Using data augmentation on different datasets with different parameters will produce different results)

I just re-ran the [Colab](https://colab.sandbox.google.com/drive/1GHFhKEzyq5fReAA05MqUn7__mU2EDaiF#scrollTo=jJhKAMhMI2D6) with data augmentation disabled and got 0.97265625 on test set and 0.979321 on train set (i.e. slightly higher on train set, as expected).

Where exactly did you modify `CIFAR_MEAN` and `CIFAR_STD` ?

**Comment by chaoyanghe** [comment]
Mine code is based on my own PyTorch Implementation. I fixed the issue of low training accuracy by changing the transforms from

    transform_train = transforms.Compose([
        transforms.RandomSizedCrop((args.img_size, args.img_size), scale=(0.05, 1.0)),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        transforms.Normalize(CIFAR_MEAN, CIFAR_STD),
    ])


to 
    transform_train = transforms.Compose([
        transforms.Resize(args.img_size),
        transforms.RandomCrop(args.img_size),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        transforms.Normalize(CIFAR_MEAN, CIFAR_STD),
    ])

This confirms that the issue belongs to "RandomSizedCrop".


Also the 0.5 normalization can increase the accuracy around 3%.

But now I still onl get 96.5% test accuracy, cannot replicate the 98% test accuracy. May I know what other data augmentation or tricks should I use?

**Comment by andsteing** [comment]
Indeed, I think in your snippet the observed difference has more to due with `RandomSizedCrop()` than with `Normalize()`.

It might be interesting to see if this observed difference is larger for ViT than for ResNet/MobileNet, but I would expect for ResNet/MobileNet also to have a deterioration in training accuracy if you apply `RandomSizedCrop()` with the same parameters.

**Comment by chaoyanghe** [comment]
Could you help to check my code? It is a single file that expresses the training and tricks.
https://github.com/FedML-AI/FedML/blob/fed-transformer/fedml_experiments/centralized/fed_transformer/main_vit.py

I am not sure whether mine reimplementation is correct or not.

**Comment by chaoyanghe** [comment]
> Indeed, I think in your snippet the observed difference has more to due with `RandomSizedCrop()` than with `Normalize()`.
> 
> It might be interesting to see if this observed difference is larger for ViT than for ResNet/MobileNet, but I would expect for ResNet/MobileNet also to have a deterioration in training accuracy if you apply `RandomSizedCrop()` with the same parameters.

I never experienced such a phenomenon before. I don't think ResNet has such an issue.
